File: 023_merge_k_sorted.ts
class ListNode {
    val: number;
    next: ListNode | null;
    constructor(val?: number, next?: ListNode | null) {
        this.val = val === undefined ? 0 : val;
        this.next = next === undefined ? null : next;
    }
}

function mergeTwoLists(
    list1: ListNode | null,
    list2: ListNode | null
): ListNode | null {
    const dummy1 = new ListNode(0, list1);
    let prev: ListNode | null = dummy1;

    while (list1 !== null && list2 !== null) {
        if (list1.val <= list2.val) {
            list1 = list1.next;
            prev = prev!.next;
        } else {
            let temp = list2.next;
            prev!.next = list2;
            list2.next = list1;
            list2 = temp;
            list1 = prev!.next;
        }
    }

    while (list2) {
        prev!.next = list2;
        prev = prev!.next;
        list2 = list2.next;
    }

    return dummy1.next;
}

function printLL(head: ListNode | null): void {
    let curr = head;
    const arr: number[] = [];
    while (curr) {
        arr.push(curr.val);
        curr = curr.next;
    }
    console.log(arr);
}

function mergeKLists(lists: Array<ListNode | null>): ListNode | null {
    if (lists.length === 0) return null;

    while (lists.length > 1) {
        const temp: Array<ListNode | null> = [];
        for (let i = 0; i < lists.length; i += 2) {
            temp.push(
                mergeTwoLists(
                    lists[i],
                    i + 1 < lists.length ? lists[i + 1] : null
                )
            );
        }
        lists = temp;
    }

    return lists[0];
}

File: 12thJune2024/023_merge_k_sorted.cpp
#include <bits/stdc++.h>
using namespace std;

struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
  int count = 0;

  ListNode *mergeTwoLists(ListNode *list1, ListNode *list2) {
    ListNode *dummy = new ListNode();
    ListNode *tail = dummy;

    while (list1 && list2) {
      count++;
      if (list1->val < list2->val) {
        tail->next = list1;
        list1 = list1->next;
      } else {
        tail->next = list2;
        list2 = list2->next;
      }
      tail = tail->next;
    }

    if (list1) {
      tail->next = list1;
    }

    if (list2) {
      tail->next = list2;
    }

    return dummy->next;
  }

public:
  ListNode *mergeKLists(vector<ListNode *> &lists) {

    if (lists.size() == 0)
      return nullptr;

    ListNode *ans = lists[0];

    for (int i = 1; i < lists.size(); i++) {
      ans = mergeTwoLists(ans, lists[i]);
    }

    return ans;
  }
};
